### ISSUE TITLE
Closed caption support in video_player

### ISSUE THREAD

**Issue by localpcguy** [flutter/flutter#25388]
The `video_player` plugin does not seem to support Closed Captions. It is generally something that a video plugin should support for accessibility reasons. It is also a requirement for the app we are currently building in Flutter.

- Is support for closed captions something that is on the near (next month or two) roadmap for the `video_player` plugin? 
- If not, would a PR adding that be considered for merging?

**Comment by bach942** [comment]
I would be great to have this feature. We've done both the CC in both ExoPlayer and AVPlayer at work so it would be great to have a flutter plugin support and not have to write a platform wrapper.

**Comment by cbenhagen** [comment]
I think this might be too high-level for the VideoPlayer plugin as you would also need some controls for it and a way to style or position the captions. It might be better to add this to a higher level plugin like [Chewie](https://github.com/brianegan/chewie). There is on open issue for this [here](https://github.com/brianegan/chewie/issues/77).

**Comment by localpcguy** [comment]
If it isn't built into VideoPlayer, it puts the onus of accessibility on the individual teams to implement. If Flutter's goal is to be an accessible platform by default, then I still maintain it is core functionality. 

That said, a separate plugin would be better than relying on a specific UI plugin implementing it. But I don't personally believe it's ideal from an accessibility perspective.

**Comment by bach942** [comment]
I would agree with @localpcguy is saying. That since flutter is pushing accessibility, it should be built into the core video player and not an add on plugin. It may speed up the development of the core for the video player development, but at the expense of the implementing developers.

**Comment by localpcguy** [comment]
I also wanted to add that if it isn't in core, it is, IMO, less likely to be implemented. Even if there is a plugin available, it's just another friction point and will lead to lower accessibility in videos implemented with the VideoPlayer plugin.

**Comment by msal4** [comment]
Is there any update on this one?

**Comment by snaeji** [comment]
> Is there any update on this one?

Here is the issue for flutter:
[https://github.com/brianegan/chewie/issues/77](https://github.com/brianegan/chewie/issues/77)

**Comment by snaeji** [comment]
It's in the p2 goals so hopefully soon....

**Comment by JohnGalt1717** [comment]
Definately needs to be in the player because both native Oses (and future oses too) support captioning in their video players. That's where it should belong.

**Comment by vanlooverenkoen** [comment]
I would like to see it in the core package as well. Any update?

**Comment by rakyat-game** [comment]
Minimal update only support SubRip(.srt) type

Update to core plugins video_player two files:
- android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayerPlugin.java
- lib/video_player.dart

For displaying please look at 
[This](https://github.com/brianegan/chewie/issues/77)

File Attachment
[VideoPlayerPlugin.zip](https://github.com/flutter/flutter/files/3300435/VideoPlayerPlugin.zip)
[video_player.zip](https://github.com/flutter/flutter/files/3300441/video_player.zip)

**Comment by adar2378** [comment]
@rakyat-game  you should ask them to merge your code if it is stable

**Comment by localpcguy** [comment]
As part of this story I think it should handle having the close captions as part of the video stream, not just as a separate SRT file.

**Comment by JohnGalt1717** [comment]
@localpcguy That would be great but I wouldn't hold up this scenario getting to production for that work to be completed. The video player needs to be iterated early and often right now because it is missing so many basic features (i.e. DRM)

**Comment by localpcguy** [comment]
I agree, some support is better than none! I should have clarified that I'm excited to see progress on this issue. I was just noting that so either this issue can stay open (or if this is closed a new issue opened) to track embedded close caption support as well.

**Comment by snaeji** [comment]
I was able to build on @rakyat-game code and read VTT files and default to subtitles in the HLS stream.

The solution is rather dirty, language default is hardcoded and defaults to Icelandic subtitles if they are included in the HLS stream.
But this will help somebody if someone wants to use it for reference etc.

https://github.com/snaeji/video_player-0.10.1-3
https://github.com/snaeji/chewie-0.9.7

**Comment by Frank-CG** [comment]
@snaeji , thanks for the solution of https://github.com/snaeji/video_player-0.10.1-3, it works fine for Android, but it seems doesn't support for iOS platform, is there any solution for this?

**Comment by snaeji** [comment]
Hey @Frank-CG , I'm sorry that i didn't mention it but this solution is for Android only. I know that you can sideload subtitles on IOS and of course display subtitles from hls but not sure how you would fetch the data and send it over the video_player plugin.

**Comment by Frank-CG** [comment]
@snaeji , thanks for your recommendation, I want to display closed caption that embedded in hls streams, I'm trying to work on it.

**Comment by snaeji** [comment]
Good luck! Wish i had more knowledge on the iOS side but no help from me.

**Comment by Joran-Dob** [comment]
Is there an update on this? This is a critical feature of our app.

**Comment by localpcguy** [comment]
@snaeji We have implemented CCs before natively in both iOS and Android and could potentially provide support in the form of answering questions or pointing in the right direction for iOS. We've been monitoring this issue and may even try to contribute a solution once we can find the time (working at a startup, so super busy right now just getting the basics done).

**Comment by Joran-Dob** [comment]
I now created my own package for this implementation in the dart, it wraps a video player and accepts a video player in its controller to now the current position. 

You can find it [here](https://pub.dev/packages/subtitle_wrapper_package), hope it helps

**Comment by rlee1990** [comment]
Any updates this should be set to a critical level without CC a big amount of customers are left out.

**Comment by johnsonmh** [comment]
I'm going to look into this a little today. Can't promise anything, but it might be interesting to have...

* Another package for parsing common subtitle file types into dart objects.
* The `video_player` package could depend on that package and add support for associating a parsed subtitle file to the `VideoPlayerController`. 
*`VideoPlayerController` could have a field `caption`, that gives the caption for the current `position` (or `null` if there is no caption for the current timestamp).
* `video_player` could have a basic widget for displaying CC that would take in a `VideoPlayerController`, and read the `caption` field. Although there would be no requirement that users use this Widget because other libraries could display CC however they want by reading `VideoPlayerController.caption`.

**Comment by abdoutech93** [comment]
This #2473 Pull Request support hls, I mean if this closed caption will be get from hls file ? When that will be merged ?

**Comment by rlee1990** [comment]
Any update on when this will be released?

**Comment by johnsonmh** [comment]
https://github.com/flutter/plugins/pull/2487 has landed and version `0.10.7` is live on [pub.dev](https://pub.dev/packages/video_player/versions/0.10.7)! These changes include support for supplying a `ClosedCaptionFile` to the `VideoPlayerController`, and reading `VideoPlayerValue.caption` in your app. See the [example app](https://github.com/flutter/plugins/blob/9859a383c7d15966eb9165f2616afeb4ee53129a/packages/video_player/video_player/example/lib/main.dart#L208) for a basic example of how to use closed captions in your app. 

Right now the only file type that is supported out of the box is [SubRip](https://en.wikipedia.org/wiki/SubRip), but anyone can implement and use `ClosedCaptionFile`  with `VideoPlayerController`! 

I'm going to close this issue now that basic support exists, any future bugs or feature requests (such as supporting additional file types) should be filed separately.

**Comment by JohnGalt1717** [comment]
You know that WebVTT is THE standard right? HTML5 standard. Everything online is going to come with WebVTT. Hackers love subrip but it isn't the standard.

**Comment by johnsonmh** [comment]
@JohnGalt1717 Happy to look at supporting file types that people would find useful. Please file your feature request (or pr!) separately. You can reference this issue in subsequent issues or pull requests.

**Comment by snaeji** [comment]
I agree with @JohnGalt1717 that we should not close this issue since it's a major accessibility concern. When live streaming HLS for example the subtitles are embedded in the stream.

**Comment by johnsonmh** [comment]
Thanks for the feedback, I've filed https://github.com/flutter/flutter/issues/50595 to track this issue, please add as much detail as possible (and add thumbs up) to that issue. Specifically, it would be helpful if you could give an example of a HLS that demonstrates what type use case you want to have supported.
